crates/engine/src/host_component.rs

```diff
@@ -0,0 +1,89 @@
+use std::{any::Any, marker::PhantomData};
+
+use spin_manifest::CoreComponent;
+use wasmtime::Linker;
+
+use crate::RuntimeContext;
+
+/// Represents a host implementation of a Wasm interface.
+pub trait HostComponent: Send + Sync {
+    /// Host component runtime state.
+    type Data: Any + Send;
+
+    /// Add this component to the given Linker, using the given runtime state-getting closure.
+    fn add_to_linker<T>(
+        linker: &mut Linker<RuntimeContext<T>>,
+        data_handle: HostComponentsDataHandle<Self::Data>,
+    ) -> anyhow::Result<()>;
+
+    /// Build a new runtime state object for the given component.
+    fn build_data(&self, component: &CoreComponent) -> anyhow::Result<Self::Data>;
```

Comment:
`data` is such an open-ended term - I found it hard to understand what this member did at the point of implementation.  Can we be more specific e.g. `build_runtime_state`?

Reply:
`data` is the term used by `wit-bindgen` to refer to the runtime state, so the name does make sense.